src/codegen/compile.ml

```diff
@@ -3805,147 +3805,63 @@ module StableMem = struct
           E.else_trap_with env "StableMemory range out of bounds")
     | _ -> assert false
 
-  (* read word32 from stable mem offset on stack *)
-  let read_word32 env =
+  let read env name typ bytes load =
     match E.mode env with
     | Flags.ICMode | Flags.RefMode ->
-      Func.share_code1 env "__stablemem_read_word32"
-        ("offset", I32Type) [I32Type]
+      Func.share_code1 env ("__stablemem_read_" ^ name)
+        ("offset", I32Type) [typ]
         (fun env get_offset ->
-          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
-            get_temp_ptr ^^ get_offset ^^  compile_unboxed_const 4l ^^
+          let words = Int32.div (Int32.add bytes 3l) 4l in
+          Stack.with_words env "temp_ptr" words (fun get_temp_ptr ->
+            get_temp_ptr ^^ get_offset ^^  compile_unboxed_const bytes ^^
             IC.system_call env "ic0" "stable_read" ^^
-            get_temp_ptr ^^ load_unskewed_ptr))
+            get_temp_ptr ^^ load))
     | _ -> assert false
 
-  (* write stable mem offset and value, operands on stack *)
-  let write_word32 env =
+  let write env name typ bytes store =
     match E.mode env with
     | Flags.ICMode | Flags.RefMode ->
-      Func.share_code2 env "__stablemem_write_word32"
-        (("offset", I32Type),("value", I32Type)) []
+      Func.share_code2 env ("__stablemem_write_" ^ name)
+        (("offset", I32Type), ("value", typ)) []
         (fun env get_offset get_value ->
-          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
-            get_temp_ptr ^^ get_value ^^ store_unskewed_ptr ^^
+          let words = Int32.div (Int32.add bytes 3l) 4l in
+          Stack.with_words env "temp_ptr" words (fun get_temp_ptr ->
+            get_temp_ptr ^^ get_value ^^ store ^^
             get_offset ^^
-            get_temp_ptr ^^ compile_unboxed_const 4l ^^
+            get_temp_ptr ^^ compile_unboxed_const bytes ^^
             IC.system_call env "ic0" "stable_write"))
     | _ -> assert false
 
-  (* read word8 from stable mem offset on stack *)
-  let read_word8 env =
-    match E.mode env with
-    | Flags.ICMode | Flags.RefMode ->
-      Func.share_code1 env "__stablemem_read_word8"
-        ("offset", I32Type) [I32Type]
-        (fun env get_offset ->
-          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
-            get_temp_ptr ^^ get_offset ^^ compile_unboxed_const 1l ^^
-            IC.system_call env "ic0" "stable_read" ^^
-            get_temp_ptr ^^
-            G.i (Load {ty = I32Type; align = 0; offset = 0l; sz = Some Wasm.Types.(Pack8, ZX)})))
-    | _ -> assert false
+  let read_word32 env =
+    read env "word32" I32Type 4l load_unskewed_ptr
```

Comment:
The common usage pattern is `read env "thingy4xN" ⟨wasmtype⟩ N ⟨dest⟩`, you can exploit that if you want...

Reply:
Life is too short man. 

Was wondering about about pulling the range_check in to avoid the nested function call...